aea/cli/generate_all_protocols.py

```diff
@@ -0,0 +1,594 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2021-2022 Valory AG
+#   Copyright 2018-2019 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""
+Generate all the protocols from their specifications.
+
+This script takes all the protocol specification (scraped from the protocol README)
+and calls the `aea generate protocol` command.
+
+Currently, it does a lot of assumptions, and might not be useful for
+all use cases. However, with not much work, can be customized to achieve
+the desired outcomes.
+
+It requires the `aea` package, `black` and `isort` tools.
+"""
+import logging
+import os
+import pprint
+import re
+import shutil
+import subprocess  # nosec
+import sys
+import tempfile
+from io import StringIO
+from itertools import chain
+from operator import methodcaller
+from pathlib import Path
+from typing import Any, Iterator, List, Match, Optional, Tuple, cast
+
+import click
+import semver
+
+from aea.cli.registry.utils import download_file, extract, request_api
+from aea.common import JSONLike
+from aea.configurations.base import ComponentType, ProtocolConfig, ProtocolSpecification
+from aea.configurations.constants import DEFAULT_PROTOCOL_CONFIG_FILE
+from aea.configurations.data_types import PackageId, PublicId
+from aea.configurations.loader import (
+    ConfigLoader,
+    ConfigLoaders,
+    load_component_configuration,
+)
+from aea.exceptions import enforce
+
+
+SPECIFICATION_REGEX = re.compile(r"(---\nname.*\.\.\.)", re.DOTALL)
+LIBPROTOC_VERSION = "libprotoc 3.19.4"
+CUSTOM_TYPE_MODULE_NAME = "custom_types.py"
+README_FILENAME = "README.md"
+PROTOCOLS_PLURALS = "protocols"
+PROTOCOL_GENERATOR_DOCSTRING_REGEX = "It was created with protocol buffer compiler version `libprotoc .*` and aea version `.*`."
+DEFAULT_TEST_SAMPLES = [
+    ("sample_specification.yaml", "t_protocol",),
+    ("sample_specification_no_custom_types.yaml", "t_protocol_no_ct",),
+]
+
+
+logging.basicConfig(format="[%(asctime)s][%(levelname)s] %(message)s")
+logger = logging.getLogger("generate_all_protocols")
+logger.setLevel(logging.INFO)
+
+
+def check_working_tree_is_dirty() -> None:
```

Comment:
Generic utility. Check if used elsewhere. In any case, make it available in a generic utils location so its reused everywhere and easily found.